### PR TITLE
Feature/resiliency

### PR DESCRIPTION
**Pull request by adarsh-dell**
# Description
End points for nodeProbe.

# GitHub Issues
List the GitHub issues impacted by this PR:

| GitHub Issue # |
| -------------- |
|https://github.com/dell/csm/issues/587|

# Checklist:

- [x] I have performed a self-review of my own code to ensure there are no formatting, vetting, linting, or security issues
- [x] I have verified that new and existing unit tests pass locally with my changes
- [x] I have not allowed coverage numbers to degenerate
- [ ] I have maintained at least 90% code coverage
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] Backward compatibility is not broken

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Please also list any relevant details for your test configuration

- [x] Called all endpoints by using cURL command and got expected http output/response.

### CODE REVIEW

File: helm/csi-powerstore/templates/controller.yaml [hunk 1]
```diff
@@ -360,6 +360,22 @@ spec:
...
+            {{- if hasKey .Values.node "podmon" }}
+            - name: X_CSI_PODMON_ENABLED
+              value: "{{ .Values.podmon.enabled }}"
+            {{- if eq .Values.podmon.enabled true }}
+              {{- range $key, $value := .Values.podmon.node.args }}
+                {{- if contains "--arrayConnectivityPollRate" $value }}
+            - name: X_CSI_PODMON_ARRAY_CONNECTIVITY_POLL_RATE
+              value: "{{ (split "=" $value)._1 }}"
+                {{ end }}
+              {{ end }}
```
Comment (spriya-m) [inline helm/csi-powerstore/templates/controller.yaml:372]: In lines 371, 372, 373: "- end" or just "end" ?

File: helm/csi-powerstore/templates/node.yaml [hunk 2]
```diff
@@ -158,6 +158,22 @@ spec:
...
+            {{- if hasKey .Values.node "podmon" }}
+            - name: X_CSI_PODMON_ENABLED
+              value: "{{ .Values.podmon.enabled }}"
+            {{- if eq .Values.podmon.enabled true }}
+              {{- range $key, $value := .Values.podmon.node.args }}
+                {{- if contains "--arrayConnectivityPollRate" $value }}
+            - name: X_CSI_PODMON_ARRAY_CONNECTIVITY_POLL_RATE
+              value: "{{ (split "=" $value)._1 }}"
+                {{ end }}
+              {{ end }}
```
Comment (spriya-m) [inline helm/csi-powerstore/templates/node.yaml:170]: Lines 169, 170, 171: "-end" or just "end" ?

File: pkg/node/node.go [hunk 3]
```diff
@@ -170,6 +174,18 @@ func (s *Service) Init() error {
 		s.isHealthMonitorEnabled, _ = strconv.ParseBool(isHealthMonitorEnabled)
 	}
 
+	if isPodmonEnabled, ok := csictx.LookupEnv(ctx, common.EnvPodmonEnabled); ok {
+		// in case of any error in reading/parsing the env variable default value will be false
+		s.isPodmonEnabled, _ = strconv.ParseBool(isPodmonEnabled)
+		/* // EnvPodmonArrayConnectivityPollRate
+		if s.isPodmonEnabled {
```
Comment (spriya-m) [inline pkg/node/node.go:181]: Please remove commented code. Please check other files as well.

File: pkg/node/node_connectivity_checker.go [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+	_, err = w.Write(jsonResponse)
+	if err != nil {
+		w.WriteHeader(http.StatusOK)
+		w.Header().Set("Content-Type", "application/json")
+		log.Errorf("unable to write response %s", err)
+	}
+}
+
+// MarshalSyncMapToJSON marshal the sync Map to Json
+func MarshalSyncMapToJSON(m *sync.Map) ([]byte, error) {
```
Comment (spriya-m) [inline pkg/node/node_connectivity_checker.go:144]: Suggestion for future improvements: If these functions are duplicated in each driver code to run a HTTP server and handle HTTP requests, we can move these common functions to a seperate library.

File: helm/csi-powerstore/templates/node.yaml (hunk 2 above)
Comment (adarsh-dell) [reply helm/csi-powerstore/templates/node.yaml]: Done

File: pkg/node/node_connectivity_checker.go (hunk 4 above)
Comment (adarsh-dell) [reply pkg/node/node_connectivity_checker.go:144]: Sure Priya. Nice idea.

**Review by AkshaySainiDell — DISMISSED** [review]

File: pkg/common/envvars.go [hunk 5]
```diff
@@ -98,4 +98,13 @@ const (
 
 	// EnvAllowAutoRoundOffFilesystemSize specifies if auto round off minimum filesystem size is enabled
 	EnvAllowAutoRoundOffFilesystemSize = "CSI_AUTO_ROUND_OFF_FILESYSTEM_SIZE"
+
+	//EnvPodmonEnabled indicates that podmon is enabled
+	EnvPodmonEnabled = "X_CSI_PODMON_ENABLED"
+
+	//EnvPodmonAPIPORT indicates the port to be used for exposing podmon API health, TO-DO - Is it okay to rename this var that makes more sense. Ex: EnvPodmonArrayConnectivityAPIPORT
```
Comment (suryagupta4) [inline pkg/common/envvars.go:105]: Do we need to take care of this To-Do item mentioned here in the comment?

Comment (adarsh-dell) [reply pkg/common/envvars.go:105]: Not for as of now. In case if other team will rename or will agree to rename, we will also do the same.

**Review by suryagupta4 — APPROVED** [review]

**Review by suryagupta4 — APPROVED** [review]

**Review by AkshaySainiDell — APPROVED** [review]

**Pull request merged**